### ISSUE TITLE
config/* files can't be kustomized as-is w/o envsubst

### ISSUE THREAD

**Issue by jdef** [kubernetes-sigs/cluster-api#3881]
In recent times it seems that feature-gate flags were added to CAPI controllers. The values for these flags seem to be assigned by envsubst variable expansion e.g. `${EXP_SOMETHING:=false}`. This convention renders the config/* manifests not usable as-is using *only* kustomize, and requires an additional tool, envsubst, in order to properly generate kustomized manifests.

I'd like for CAPI to eliminate this additional binary requirement of envsubst. It seems that in the CAPI codebase that the feature gate flags are used by dev and CI. I'm hoping there's another way to accomplish the same thing in order to eliminate the dependency on envsubst.

A somewhat related issue is #3252, but that has more to do with cluster templates.

/cc @fabriziopandini

**Comment by fabriziopandini** [comment]
@jdef thanks for this report
IMO fixing this is in the scope of the work of the management cluster operator, given that it is redesigning how the user can configure providers in a fully declarative way. I have added a comment on the proposal to make https://github.com/kubernetes-sigs/cluster-api/pull/3833/files#r512598988 to make sure this is in the radar
cc @wfernandes 

WRT to the current problem, 

> This convention renders the config/* manifests not usable as-is using only kustomize, and requires an additional tool...

I know that kustomize can still be used over config/* because we are still doing in several places (config/e2e);
I'm wondering if, as a temporary workaround, you can add a patch that replaces the lines where ${EXP_SOMETHING:=false} is used with the actual values. Not ideal, but it should work...

**Comment by jdef** [comment]
> I know that kustomize can still be used over config/* 

To be clear, my expectation is that one should be able to successfully execute a pipeline along the lines of:
```
:; kustomize build config | kubectl apply -f -
```
.. and this is currently not possible given the additional envsubst (or similar) requirement.

**Comment by fabriziopandini** [comment]
gotcha, this is where the additional patch could help (to remove the need of envsubst)

**Comment by benmoss** [comment]
I think the problem with this ask is that it would mean that this feature flags (`EXP_MACHINE_POOL`, etc) would be set at manifest creation time, whereas with the flow we document in the quick-start tutorial we support turning these features on at deploy time.

One way to get around that might be to have multiple kustomize overlays, one of which is the deferred/envsubst version, but others of which resolve to fully finalized yaml. I think we'll still get into the problem with kustomize of needing one overlay per feature, or an overlay that is "all experiments turned on" or something.

**Comment by jdef** [comment]
kustomize recently added "components" (kind of like patchset mixins). maybe that could help? https://github.com/kubernetes-sigs/kustomize/blob/master/examples/components.md

**Comment by benmoss** [comment]
Yeah I was looking at this, I think we could do a component like `machinepools` that inserts the `--feature-gates=MachinePool=true` argument onto the deployment, but then you still have to incorporate that component into an overlay, so you end up in the position of needing 1 overlay per component, at least if you want to give your users that same flexibility of turning on each component individually.

**Comment by fabriziopandini** [comment]
Summary of a recent discussion about this:
- the management cluster operator work is going to remove the need of variables for feature flags
- the work on multitenancy could potentially remove the need of variables for credentials (to be confirmed)
- at least in Azure, there is still the need for env vars for AZURE_ENVIRONMENT

**Comment by CecileRobertMichon** [comment]
> at least in Azure, there is still the need for env vars for AZURE_ENVIRONMENT

@nader-ziada @devigned wdyt about moving `AZURE_ENVIRONMENT` to AzureCluster spec so it's no longer needed in the CAPI yaml? If subID/credentials are per cluster, there's no reason Azure Cloud shouldn't also be per cluster.

**Comment by fabriziopandini** [comment]
/triage accepted
/help-wanted

we have to re-assess the situation and see if we can remove all the variables from those files as suggested above

**Comment by fabriziopandini** [comment]
before acting we should reconcile requirements with https://github.com/kubernetes-sigs/cluster-api/issues/3247 that advocates for more templating features into provider yaml files

**Comment by fabriziopandini** [comment]
/lifecycle frozen
/help

even if there are conflicting requirements between this issue and https://github.com/kubernetes-sigs/cluster-api/issues/3247, finding a way forward could help people using Cluster API without cluterctl
